File: contest/typical90AK/typical90AK.cpp

#include <bits/stdc++.h>
#include <atcoder/segtree>
using namespace std;
using namespace atcoder;

using ll = long long;
ll W, N;

struct S {
    ll a;
};

S op(S l, S r) { return S{max(l.a, r.a)}; }

S e() { return S{-1}; }


int main(){
    cin >> W >> N;
    vector<S> a(W+1, e());
    a[0] = S{0};
    vector<segtree<S, op, e>> dp(N+1, segtree<S, op, e>(a));

    for (int i=1; i<=N; i++) {
        ll l, r, v;
        cin >> l >> r >> v;
        for (int j=1; j<=W; j++) {
            ll pv = dp[i-1].prod(max(0LL, j-r), max(0LL, j-l+1)).a;
            ll now = dp[i].get(j).a;
            ll nex = max(now, dp[i-1].get(j).a);
            if (pv != -1) nex = max(nex, pv + v);
            dp[i].set(j, S{nex});
        }
    }
    cout << dp[N].get(W).a << endl;

    return 0;
}

File: contest/typical90AK/typical90AK.py
import sys

input = lambda: sys.stdin.readline().rstrip()


class SegTree:
    def __init__(self, num, e_factory, operand=max):
        """
        parameter
            num : 要素数
            e_factory : 単位元を生成する関数
            operand : 子ノードの値から親ノードの値を作る関数
        """
        # self.tree[i+self.offset]:要素iのみの区間の値
        self.offset = 1 << (num - 1).bit_length()
        self.tree = [e_factory() for _ in range(self.offset << 1)]
        self.e_factory = e_factory
        self.operand = operand

    def update(self, i, value):  # O(logN)
        """i番目の要素をvalueにする"""
        i += self.offset
        self.tree[i] = value
        # 親ノードを順に変更していく
        while i > 1:
            v1, v2 = self.tree[i], self.tree[i ^ 1]  # 子ノード
            i >>= 1  # iを親ノードへずらす
            self.tree[i] = self.operand(v1, v2)

    def get_range(self, l, r):  # O(logN)
        """区間[l, r)の値を取得する"""
        l += self.offset
        r += self.offset

        ans = self.e_factory()
        while l < r:
            if r & 1:
                ans = self.operand(ans, self.tree[r - 1])
                r -= 1
            if l & 1:
                ans = self.operand(self.tree[l], ans)
                l += 1
            l >>= 1
            r >>= 1
        return ans


W, N = map(int, input().split())

dp = [SegTree(W + 1, e_factory=lambda: -1, operand=max) for _ in range(N + 1)]
for i in range(N + 1):
    dp[i].update(0, 0)
# dp[i][j]: i番目まで見て、香辛料をちょうどj使った時の価値の最大値
for i in range(1, N + 1):
    l, r, v = map(int, input().split())
    for j in range(1, W + 1):
        pv = dp[i - 1].get_range(max(0, j - r), max(0, j - l + 1))
        now = dp[i].get_range(j, j + 1)
        nex = max(now, dp[i - 1].get_range(j, j + 1))
        if pv != -1:
            nex = max(nex, pv + v)
        dp[i].update(j, nex)
    # print(dp[i].tree[dp[i].offset :])

ans = dp[N].get_range(W, W + 1)
print(ans)
